### ISSUE TITLE
Performant way to create stateful proxies

### ISSUE THREAD

**Issue by RainerGanss** [raphw/byte-buddy#1199]
Hi,

TL;DR;
I have millions of instances of ~400 classes and want to create stateful proxies for each one. What is the best way?

My use case: I have a big model in two version which are incompatible. I wrote a compatibility lib which overcomes those incompatibilities with cglib and it worked just fine. Now due to java 17 migration I cannot use cglib and would like to use byte buddy. 

Replacing the old code 
```java
final Enhancer e = new Enhancer();
e.setSuperclass(mappedClass);
e.setCallback(callback); //callback is stateful and implements the cglib MethodInterceptor
return (T) e.create();
```
was somewhat easy:
```java
final Class<?> proxyType = new ByteBuddy()
            .subclass(mappedClass)
            .method(not(isDeclaredBy(Object.class)))
            .intercept(to(callback)) //callback is stateful and has an @RuntimeType intercept method
            .make()
            .load(target.getClass().getClassLoader())
            .getLoaded());
return (T) proxyType.newInstance();
```
This works but is super slow. Runtime 2 min vs 2 sec.  

What is the correct way to do this? Ideally I would like to separate the expensive proxy creation from setting the interceptor but from what I understand that is not possible due the way, bb works.

**Comment by raphw** [comment]
Have a look at TypeCache, cglib has some built-in cache with implicit keys but it has a row of weaknesses.

Instead of delegating to an instance, you delegate to a field that you have defined and then you set that field on each instance of your proxy. The class itself should be reused via the cache.

**Comment by RainerGanss** [comment]
Wow thanks for the super fast repsonse! I rewrote the code based on your tests to
```java
final Class<?> proxyType = typeCache.findOrInsert(target.getClass().getClassLoader(), target.getClass(), () ->
new ByteBuddy().subclass(mappedClass)
        .defineField(QUX, InvocationHandler.class, Visibility.PUBLIC)
        .method(not(isDeclaredBy(Object.class)))
        .intercept(InvocationHandlerAdapter.toField(QUX).withoutMethodCache())
        .make()
        .load(target.getClass().getClassLoader())
        .getLoaded());
final T proxy = (T) proxyType.newInstance();
final Field field = proxy.getClass().getDeclaredField(QUX);
field.set(proxy, callback);
return proxy;
```
And this works! So thanks for that. 

Now I only have the problem, that the InvocationHandler-Interface does not support the MethodProxy from cglib or @SuperMethod from byte buddy. This is of topic ik, but do you have a hint for me? My (years old code) requires it (and it ran with @SuperMethod).

**Comment by RainerGanss** [comment]
Ok so I worked around the need for the super method parameter by excluding a method from being handeled by the proxy at all. 

`.method(not(named("accept")).and(not(isDeclaredBy(Object.class))))`

Thx again for your help - this can be closed.

**Comment by raphw** [comment]
If you wanted such a super method call, have a look at MethodDelegation and the parameter annotations that are described in its javadoc.
